core/embed/rust/src/ui/display.rs

```diff
@@ -105,6 +151,341 @@ pub fn rect_fill_rounded1(r: Rect, fg_color: Color, bg_color: Color) {
+                center,
+                Offset::new(toif_info.width.into(), toif_info.height.into()),
+            );
+            icon_area_clamped = icon_area.clamp(constant::screen());
+
+            let mut ctx = uzlib::UzlibContext::new(&i.0[12..], false);
+            if ctx.uncompress(&mut icon_data).is_err() {
+                return;
+            }
+            icon_colortable = get_color_table(i.1, bg_color);
+            icon_width = toif_info.width.into();
+            use_icon = true;
+        }
+    }
+
+    let start = 0;
+    let end = (start + ((360 * show_percent) / 100)) % 360;
+
+    let start_vector;
+    let end_vector;
+
+    let mut show_all = false;
+    let mut inverted = false;
+
+    if show_percent >= 100 {
+        show_all = true;
+        start_vector = Point::zero();
+        end_vector = Point::zero();
+    } else if show_percent > 50 {
+        inverted = true;
+        start_vector = get_vector(end);
+        end_vector = get_vector(start);
+    } else {
+        start_vector = get_vector(start);
+        end_vector = get_vector(end);
+    }
+
+    let n_start = Point::new(-start_vector.y, start_vector.x);
+
+    for y_c in r.y0..r.y1 {
+        for x_c in r.x0..r.x1 {
+            let p = Point::new(x_c, y_c);
+
+            let mut icon_pixel = false;
+            if use_icon && icon_area_clamped.contains(p) {
+                let x_i = p.x - icon_area.x0;
+                let y_i = p.y - icon_area.y0;
+
+                let data = icon_data[(((x_i & 0xFE) + (y_i * icon_width)) / 2) as usize];
+                if (x_i & 0x01) == 0 {
+                    pixeldata(icon_colortable[(data >> 4) as usize]);
+                } else {
+                    pixeldata(icon_colortable[(data & 0xF) as usize]);
+                }
+                icon_pixel = true;
+            }
+
+            if !clamped.contains(p) || icon_pixel {
+                continue;
+            }
+
+            if !icon_pixel {
+                let y_p = -(p.y - center.y);
+                let x_p = p.x - center.x;
+
+                let vx = Point::new(x_p, y_p);
+                let n_vx = Point::new(-y_p, x_p);
+
+                let is_past_start = is_clockwise_or_equal(n_start, vx);
+                let is_before_end = is_clockwise_or_equal_inc(n_vx, end_vector);
+
+                if show_all
+                    || (!inverted && (is_past_start && is_before_end))
+                    || (inverted && !(is_past_start && is_before_end))
+                {
+                    let p_b = p - r.top_left();
+                    let c =
+                        rect_rounded2_get_pixel(p_b, r.width(), r.height(), colortable, false, 2);
+                    pixeldata(c);
+                } else {
+                    pixeldata(bg_color);
+                }
+            }
+        }
+    }
+
+    pixeldata_dirty();
+}
+
+/// Gets a color of a pixel on `p` coordinates of rounded rectangle with corner
+/// radius 2
+pub fn rect_rounded2_get_pixel(
```

Comment:
does this need to be `pub` ?

Reply:
it does not: https://github.com/trezor/trezor-firmware/commit/ce97b6452904d229e17dc2e672562804e24a71f2